pages/learn/develop/container-contracts.md

```diff
@@ -0,0 +1,90 @@
+
+Each service container can define the contracts that it enforces. If a container's contract requirements are not met, the release is not deployed to the device. You only need to define a contract for service containers that you wish to enforce a contract.
+
+## Contract specification
+
+Container contracts are defined as a `contract.yml` file and must be placed in the application structure alongside the corresponding service Dockerfile. For example, the following [multi-container application](/learn/develop/multicontainer/) defines two services named `first-service` and `second-service`, each with a corresponding `contract.yml` file.
+
+```shell
+.
+├── docker-compose.yml
+├── first-service
+│   ├── contract.yml
+│   └── Dockerfile.template
+└── second-service
+    ├── contract.yml
+    └── Dockerfile.template
+```    
+
+`contract.yml` is a YAML file containing `type`, `slug`, `name`, and `requires` keys, which define the requirements of the contract. The following example requires the device to be running a version of balenaOS greater or equal than `2.40.0` for the contract to pass.
+
+```yaml
+type: 'sw.container'
+slug: 'enforce-os-gt-240'
+name: 'Enforce hostOS Version'
+requires:
+  - type: 'sw.os'
+    version: '>=2.40.0'
+```
+
+* `type` should be set to `sw.container` for container contracts.
+* `slug` should be a unique identifier to identify the contract.
+* `name` should be a descriptive name of the contract.
+* `requires` should be a list of requirements to be enforced by the contract defined by `type` and `version` keys. Currently, `type` may be either `sw.os` for the hostOS version or `sw.supervisor` for the device Supervisor version. 
+
+__Note__ If `name` or `type` are not supplied, the build will fail. If `slug` is not provided, the build will complete, but the contract will fail validation on deployment. If `requires` is not provided, the contract will have no effect and the release will be deployed.
+
+Multiple contract requirements may be added to a single contract. The following example requires that the hostOS is greater than or equal to 2.44.0, and the Supervisor version is greater than 10.
+
+```yaml
+type: 'sw.container'
+slug: 'enforce-supervisor-and-os'
+name: 'Enforce hostOS and supervisor'
+requires:
+  - type: 'sw.supervisor'
+    version: '>10'
+  - type: 'sw.os'
+    version: '>=2.44.0'  
+```
+
+__Note__ Types in the contract requirements are not validated, so passing values for `type` other than `sw.os` or `sw.supervisor` will result in the contract passing.
```

Comment:
I think this is actually the opposite result, they're not checking, but there's no way that the contract engine could fulfill the contract, so it's denied.

Reply:
I struggled with the language here. The reason I added this was I could imagine bugs with typos and the contract not behaving as expected. As in, you may expect the contract to validate that types are being checked and so passing an invalid type isn't allowed whereas actually it's just going to be ignored. I'll try and clean up the language though.

Reply:
@CameronDiver actually I'm confused on this. If I pass an invalid type say `sw.oss` the contract passes. I'm struggling with wording as not sure it 'passes' in as much as it isn't enforced. Is this the intended behaviour as I am not sure 'denied' explains it any better.

```
__Note__ Types in the contract requirements are not validated, so passing values for `type` other than `sw.os` or `sw.supervisor` will result in the contract not being enforced.
```

Reply:
Yeah it's how you describe, because we're not validating types (maybe we should be in the future?) the engine will just accept any type and try to reconcile it. Because there won't be an item in the engine universe which will conform to, say `sw.oss`, the engine will deny the release.

Basically, think about it like the engine not reporting anything different when there's no contracts which fulfill a type, and there's contracts which fulfill the type, but not the version requirement. Does this help?

Reply:
Right, that's how I understand except I think the engine doesn't "deny the release" it "passes" and is deployed. I think that's the source of the confusion. Are you saying that if I set type to `sw.oss` the contract should fail and not be deployed?

Reply:
Oh, this is weird, I haven't actually tested this specific case so maybe I'm wrong. If that is indeed the case, we'll need to do validation on the input types. I'll test this and get back to you.

Reply:
I'll also confirm this with testing as I'm not so sure anymore, perhaps it wasn't deployed but there was nothing in the logs. Either way, I'll double-check.

Reply:
Resolved via https://www.flowdock.com/app/rulemotion/i-docs/threads/7nQx7ciXHAw9MYpAJZxg6nEanK6 and #1218 314b6c5bc3d670e43bad95a0f93ad09e496ac9f9